p2p/peerexchange/bookchecker.go

```diff
@@ -0,0 +1,133 @@
+package peerexchange
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// CheckBook periodically checks the book for dead|alive peers.
+func (d *Discovery) CheckBook(ctx context.Context) {
+	if d.cfg.CheckInterval == 0 {
+		d.logger.Info("periodically peers checking disabled")
+		return
+	}
+	ticker := time.NewTicker(d.cfg.CheckInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			d.logger.Debug("checking book")
+			d.CheckPeers(ctx)
+			d.logger.Debug("checking book done")
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// CheckPeers periodically checks the book for dead|alive peers.
+func (d *Discovery) CheckPeers(ctx context.Context) {
+	peers := d.GetRandomPeers(d.cfg.CheckPeersNumber)
+	if len(peers) == 0 {
+		d.logger.Info("no peers to check")
+		return
+	}
+	qCtx, cancel := context.WithTimeout(ctx, d.cfg.CheckTimeout)
+	defer cancel()
+	if _, err := d.crawl.query(qCtx, peers); err != nil {
+		d.logger.Error("failed to check nodes: %s", err)
+		return
+	}
+	// check peers are updated in host peerBook.
+	for peerID, addresses := range d.peersToMap(peers) {
+		if err := d.host.Network().ClosePeer(peerID); err != nil { // close connection, used only for check
+			d.logger.Error("failed to close peer after check: %s", err)
+		}
+		peerStoreAddresses := d.host.Peerstore().Addrs(peerID)
+		if len(peerStoreAddresses) == 0 {
+			d.book.RemoveAddress(peerID)
+			continue
+		}
+		// in case if there are more than one address for peerID is available - for us this is not important, just take first one.
+		newAddr, err := parseAddrInfo(peerStoreAddresses[0].String() + "/p2p/" + peerID.Pretty())
+		if err != nil {
+			d.logger.Error("failed to parse address: %s", err)
+			continue
+		}
+		// update node address in book.
+		d.book.AddAddress(newAddr, addresses[1])
```

Comment:
why is it addresses[1]? should it be addresses[0]? but maybe we can get rid from duplicates in our peer book instead

Reply:
yes, forgot rollback after tests. `addresses[0]` should be here

in case of multiply addresses valid after check - just take firs